test/e2e/specs/editor/various/post-editor-template-mode.spec.js

```diff
@@ -0,0 +1,371 @@
+						.locator( 'role=button[name="Cancel"i]' )
+						.click();
+				}
+
+				// Exit template mode.
+				await page.click( 'role=button[name="Back"i]' );
+
+				await pageUtils.openDocumentSettingsSidebar();
+
+				// Move focus to the "Post" panel in the editor sidebar.
+				const postPanel = postEditorTemplateMode.editorSettingsSidebar.locator(
+					'role=button[name="Post"i]'
+				);
+				await postPanel.click();
+
+				const templateSelect = postEditorTemplateMode.editorSettingsSidebar.locator(
+					'role=combobox[name="Template:"i]'
+				);
+				await expect( templateSelect ).toHaveValue(
+					`wp-custom-template-${ viewport }-viewport-deletion-test`
+				);
+			} );
+
+			test( `should delete template if deletion is confirmed when the viewport is ${ viewport }`, async ( {
+				page,
+				pageUtils,
+				postEditorTemplateMode,
+			} ) => {
+				const templateTitle = `${ viewport } Viewport Deletion Test`;
+
+				await pageUtils.setBrowserViewport( viewport );
+
+				await postEditorTemplateMode.createNewTemplate( templateTitle );
+
+				// Close the settings in small viewport.
+				if ( viewport === 'small' ) {
+					await page.click( 'role=button[name="Close settings"i]' );
+				}
+
+				// Edit the template.
+				await pageUtils.insertBlock( { name: 'core/paragraph' } );
+				await page.keyboard.type(
+					'Just a random paragraph added to the template'
+				);
+
+				await postEditorTemplateMode.saveTemplateWithoutPublishing();
+
+				{
+					const templateDropdown = postEditorTemplateMode.editorTopBar.locator(
+						'role=button[name="Template Options"i]'
+					);
+					await templateDropdown.click();
+					await page.click(
+						'role=menuitem[name="Delete template"i]'
+					);
+
+					const confirmDeletionDialog = page.locator( 'role=dialog' );
+					await expect( confirmDeletionDialog ).toBeFocused();
+					await expect(
+						confirmDeletionDialog.locator(
+							`text=Are you sure you want to delete the ${ templateTitle } template? It may be used by other pages or posts.`
+						)
+					).toBeVisible();
+
+					await confirmDeletionDialog
+						.locator( 'role=button[name="OK"i]' )
+						.click();
+				}
+
+				// Saving isn't technically necessary, but for themes without any specified templates,
+				// the removal of the Templates dropdown is delayed. A save and reload allows for this
+				// delay and prevents flakiness
+				{
+					await page.click( 'role=button[name="Save draft"i]' );
+					await page.waitForSelector(
+						'role=button[name="Dismiss this notice"] >> text=Draft saved'
+					);
+					await page.reload();
+				}
+
+				const templateOptions = postEditorTemplateMode.editorSettingsSidebar.locator(
+					'role=combobox[name="Template:"i] >> role=menuitem'
+				);
+				const availableTemplates = await templateOptions.allTextContents();
+
+				expect( availableTemplates ).not.toContain(
+					`${ viewport } Viewport Deletion Test`
+				);
+			} );
+		} );
+	} );
+} );
+
+class PostEditorTemplateMode {
```

Comment:
The only thing I wonder about is whether a class is the right option here. This seems like a set of loosely defined utility functions while a class is usually strictly modelled around a particular object. Maybe this is idiomatic playwrighting. I don't know it well enough to say for sure 😄 

It is just test code and it works well and is organised very nicely, so it's still a thumbs up 👍

Reply:
Yeah, it's debatable.

The idea is to follow the [POM](https://playwright.dev/docs/pom) convention to create utility helpers for tests for a specific page. We could just create utility functions as we did in `e2e-test-utils`, but that would mean remembering to pass all those previously global variables like `page` and `browser` to each function. Using classes allows us to implicitly bind those variables to the helpers via the `this` keyword.

Another good idea will be to just define these functions as [fixture](https://playwright.dev/docs/test-fixtures#creating-a-fixture) functions. I just default to POM to try to make them consistent 😅 . There's still room for improvements, though, so ideas are welcome!

Reply:
I see, sounds like this is a recommended approach. I think it's ok as a convention, and for anyone contributing familiar with playwright, they'll probably find it helpful that we've followed this pattern.

It also seems a bit simpler than the fixture example.

Thanks for linking to the docs. 👍